webhooks/cluster.go

```diff
@@ -62,22 +68,42 @@ func (c *Cluster) Default(_ context.Context, obj runtime.Object) error {
 }
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
-func (c *Cluster) ValidateCreate(_ context.Context, obj runtime.Object) error {
+func (c *Cluster) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	var allErrs field.ErrorList
+
 	cluster, ok := obj.(*clusterv1.Cluster)
 	if !ok {
 		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", obj))
 	}
 
+	// In the case the webhook fails the existing checks return early without waiting for ClusterTopology specific checks.
+
 	// Note: The code in ValidateCreate is intentionally not duplicated to avoid that we accidentally
 	// implement new checks in the API package and forget to duplicate them to the webhook package.
 	// The idea is to add new validation which requires a reader and Cluster variable/patch validation
 	// in the webhook package.
 	// When we drop the method in the API package we must inline it here.
-	return cluster.ValidateCreate()
+	if err := cluster.ValidateCreate(); err != nil {
+		return err
```

Comment:
I'm not happy with this mechanism of returning early but I can't see a way to:
1. Keep the old Validate method in place
2. Aggregate errors from the old Validate methods and the new checks together without some hack that would only be temporary i.e. until api.Validate* methods are fully deprecated 
@sbueringer @fabriziopandini WDYT?

Reply:
Why couldn't we aggregate the error with the rest of the function instead of returning?

Reply:
I'm not sure if it's safe if we just take the errors and aggregate, the problem is that we cannot rely on any previous validation being successful.

This might not be a problem, but it could lead to panics or that we have to check some fields twice to make sure we don't panic. (Not sure if that's the case here but we already have some duplicate nil checks in the API package because of that)

Reply:
When I think about it some more it might be interesting to implement fuzzy testing for Validation/Default funcs (they are successful if the validation doesn't panic). Of course not in this PR.

Reply:
my two cents.
- we should strive to aggregate errors because surfacing all the problems in a single pass provides a better UX to the users (vs failing, fixing, failing again for the next check, iterate)
- a check should be performed if possible (all the requires info exists, we don't panic); this IMO is more robust vs relying on the order of checks, even if it comes at the cost of some repeated nil checks

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/5594 will move all of the checks to the webhooks package so we can resolve this after it's merged.

Reply:
I think we all agree on the expected UX (don't want to return early).
The fact that accommodating the code to satisfy that UX is a bit convoluted here because we are in the middle of a deprecation shouldn't be the driving force for the resulting UX.

I think we should do what it takes in code to satisfy the UX we want. I don't see a particular problem with aggregating here? If it happens that we end up with some duplication until we finally deprecate the old method I think that's acceptable and providing a consistent and better UX worths it.

Reply:
Overall sounds good to me. I'm fine with trying to surface as much errors as possible in one validation.

I think my point for the future is probably that we should just keep on trying to structure validation function in a way that we don't have to add too much duplicate nil checks (but I think we're currently in a good state)